### PR TITLE
Handling crop render settings (dataWindowNDC)

### PR DESCRIPTION
**Pull request by dedoardo**
Uses the dataWindowNDC attribute as cropping region for coreBlackbird.

It requires this [coreBlackbird PR](https://github.com/tangent-opensource/coreBlackbird/pull/58)

Picture

### CODE REVIEW

File: plugin/hdCycles/renderParam.cpp [hunk 1]
```diff
@@ -1208,6 +1210,24 @@ HdCyclesRenderParam::_HandleFilmRenderSetting(const TfToken& key, const VtValue&
                                                                 false);
     }
 
+    if (key == "dataWindowNDC") {
```
Comment (dedoardo) [inline plugin/hdCycles/renderParam.cpp:1213]: How to access the USD tokens from Hydra?

**Comment by dedoardo** [comment]
Viewport should work fine, mplay necessitates some debugging.

File: plugin/hdCycles/renderParam.cpp (hunk 1 above)
Comment (boberfly) [reply plugin/hdCycles/renderParam.cpp:1213]: It's in the header `pxr/usd/usdRender/tokens.h`
`UsdRenderTokens->dataWindowNDC;`

**Comment by dedoardo** [comment]
With this last commit, this patch behaves like Karma both in viewport mode and when rendering with mplay.

**Comment by dedoardo** [comment]
Both tiled and non-tiled rendering now behave identically and separate the resolution of the data with that of the display.
Whenever data needs to be copied to the display, it is done with nearest point sampling. This is consistent with Karma both in the viewport and mplay.

For a `ndcDataWindow` of `(0.5, 0.1, 0.9, 0.9)`, the render pass is stretched to fit the viewport
Screenshot from 2021-05-27 11-13-35

while in mplay it preserves the aspect ratio. This is because the size of the aov is already cropped.
Screenshot from 2021-05-27 11-14-38

Are there other (or different) behavior requirements ?

**Comment by boberfly** [comment]
Hey @dedoardo 
- I haven't had a chance to see how Karma does cropped windows for the viewport, but does it also stretch like that? Seems strange but not too fussed, it's the MPlay use case which is looking like it's working great. I'd imagine the first screenshot seems more useful than stretching, but if Karma does the stretch also, then that's fine.

- As for behaviour requirements, I guess you are treating overscan as a separate task to this one?

Great work, the team will be happy!

**Comment by dedoardo** [comment]
@boberfly 
- It seemed strange to me too, but it's simple to verify. I do think having the cropping as in the original screenshot is much nicer (functional too?). It's something we can look into if there are specific requirements. I wonder if there are tools within the houdini viewport that let you select a rect as render region. If so, it should work out of the box (tm).
- Yes, overscan will come in a separate PR. I guess there is no particular rush to review/merge this for now.

**Comment by boberfly** [comment]
Hmm if it's stable for you now and it matches Karma, I think we can merge it. I'll take a look at the code after some meetings.

File: plugin/hdCycles/renderParam.cpp [hunk 2]
```diff
@@ -1261,6 +1295,19 @@ HdCyclesRenderParam::_HandleFilmRenderSetting(const TfToken& key, const VtValue&
 
     if (film_updated) {
         film->tag_update(m_cyclesScene);
+
+        // todo: Should this live in another location?
```
Comment (boberfly) [inline plugin/hdCycles/renderParam.cpp:1299]: Yeah I guess we could have our own dedicated reset buffers function. I think it's ok for now, maybe @bareya might think differently but I think lets get this running on the floor.

File: plugin/hdCycles/renderBuffer.cpp [hunk 3]
```diff
@@ -202,7 +202,7 @@ HdCyclesRenderBuffer::Clear()
     std::lock_guard<std::mutex> lock { m_mutex };
 
     size_t pixelSize = HdDataSizeOfFormat(m_format);
-    memset(&m_buffer[0], 0, m_buffer.size() * pixelSize);
+    memset(&m_buffer[0], 0, m_buffer.size());
```
Comment (boberfly) [inline plugin/hdCycles/renderBuffer.cpp:205]: Nice fix, is the pixelSize line above still needed here?

**Review by boberfly — APPROVED** [review]
Just some notes @dedoardo - just relying on you testing this out without crashes... :)

**Pull request merged**